pypowerwall/cloud/pypowerwall_cloud.py

```diff
@@ -0,0 +1,997 @@
+            if os.path.exists(self.sitefile):
+                with open(self.sitefile) as file:
+                    try:
+                        self.siteid = int(file.read())
+                    except Exception as exc:
+                        log.debug(f"Unable to determine siteid from sitefile, using ID '0' instead: {exc}")
+                        self.siteid = 0
+            else:
+                self.siteindex = 0
+        log.debug(f" -- cloud: Using site {self.siteid} for {self.email}")
+
+        # Check for auth file
+        if not os.path.exists(self.authfile):
+            msg = f"Missing auth file {self.authfile} - run setup"
+            log.warning(msg)
+            raise PyPowerwallCloudNoTeslaAuthFile(msg)
+
+    def init_poll_api_map(self) -> dict:
+        # API map for local to cloud call conversion
+        return {
+            # Somewhat Real Actions
+            "/api/devices/vitals": self.get_api_devices_vitals,
+            "/api/meters/aggregates": self.get_api_meters_aggregates,
+            "/api/operation": self.get_api_operation,
+            "/api/site_info": self.get_api_site_info,
+            "/api/site_info/site_name": self.get_api_site_info_site_name,
+            "/api/status": self.get_api_status,
+            "/api/system_status": self.get_api_system_status,
+            "/api/system_status/grid_status": self.get_api_system_status_grid_status,
+            "/api/system_status/soe": self.get_api_system_status_soe,
+            "/vitals": self.get_vitals,
+            # Possible Actions
+            "/api/login/Basic": self.api_login_basic,
+            "/api/logout": self.api_logout,
+            # Mock Actions
+            "/api/auth/toggle/supported": self.get_api_auth_toggle_supported,
+            "/api/customer": self.get_api_customer,
+            "/api/customer/registration": self.get_api_customer_registration,
+            "/api/installer": self.get_api_installer,
+            "/api/meters": self.get_api_meters,
+            "/api/meters/readings": self.get_api_unimplemented_timeout,
+            "/api/meters/site": self.get_api_meters_site,
+            "/api/meters/solar": self.get_unimplemented_api,
+            "/api/networks": self.get_api_unimplemented_timeout,
+            "/api/powerwalls": self.get_api_powerwalls,
+            "/api/site_info/grid_codes": self.get_api_unimplemented_timeout,
+            "/api/sitemaster": self.get_api_sitemaster,
+            "/api/solar_powerwall": self.get_api_solar_powerwall,
+            "/api/solars": self.get_api_solars,
+            "/api/solars/brands": self.get_api_solars_brands,
+            "/api/synchrometer/ct_voltage_references": self.get_api_synchrometer_ct_voltage_references,
+            "/api/system/update/status": self.get_api_system_update_status,
+            "/api/system_status/grid_faults": self.get_api_system_status_grid_faults,
+            "/api/troubleshooting/problems": self.get_api_troubleshooting_problems,
```

Comment:
Instead of if/elif/elif/elif/.../else we define a map with api -> function and then use it in the `self.poll()` function, making it compact and easy to extend.

Reply:
I have mixed feeling about that. I have started using that method for many of my smaller sets and find it cluttered and harder to read, especially for longer lists. I notice CoPilot prefers this.

A better option would be to move to routers (ala flask) and perhaps consider moving to asyncio / uvicorn. That would be a slightly bigger lift. :)

Reply:
Continued... To be fair... willing to entertain it.  I've been moving to this in other projects. Do we have any performance data or other indicators that this is a preferred path?

Reply:
Likely, it'll be faster to get a pointer to the required function with a key in the dictionary versus traversing all `if` and `elif`, but to me it also looks compact and more readable. As you may have noticed already I like to have small and short functions, I prefer those that fit on one screen :)

Reply:
> A better option would be to move to routers (ala flask) and perhaps consider moving to asyncio / uvicorn. That would be a slightly bigger lift. :)

Definitely, that's why we should take this big first step to decouple big functions into smaller class methods first, then we'll be able to transition into something else.

Also, once this is all reviewed and ironed out, we'll be able to plug in other tesla libraries if anything better comes into play. All we'll need to do is implement a few required methods. And then let users choose which way to interact with their powerwall/tesla gateway. But first, need to get rid of hard-coded logic here and there and make the whole thing a little bit more flexible, in my opinion.

Reply:
Fair point. Adding support for the official Tesla FleetAPI is a perfect example. Sold. ;)